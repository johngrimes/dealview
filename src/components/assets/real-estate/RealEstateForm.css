.real-estate-form {
  padding: 0 16px;
}

.form-page-inactive {
  display: none;
}
