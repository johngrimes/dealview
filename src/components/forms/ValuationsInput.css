.add-valuation-button {
  margin-top: 8px;
}

.valuations-input table .control-group input {
  border: 0;
  box-shadow: none;
  width: 100px;
  padding: 8px 16px;
}

.valuations-input table input[name^="valuations-date"] {
  width: 110px;
}

.valuations-input table input[name^="valuations-amount"] {
  width: 120px;
}

.valuations-input table input[name^="valuations-note"] {
  width: 300px;
}

.valuations-input table td {
  padding: 0;
}

.valuations-input table .control-group {
  margin: 0;
}
