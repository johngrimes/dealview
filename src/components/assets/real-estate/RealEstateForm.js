// @flow

import React from 'react'
import _ from 'lodash'
import moment from 'moment'

import InputField from 'components/forms/InputField'
import HiddenField from 'components/forms/HiddenField'
import AddressField, { AddressErrorsDefaults } from 'components/forms/AddressField'
import TextAreaField from 'components/forms/TextAreaField'
import ValuationsInput from 'components/forms/ValuationsInput/ValuationsInput'
import DateField from 'components/forms/DateField'
import * as Validations from 'utils/FormValidation'
import { RealEstateDefaults } from 'types/assets/realEstate'
import { DateStorageFormat } from 'types/commonTypes'
import { updateValuationsWithPurchaseOrSale, updateValuationsWithPurchase,
  updateValuationsWithSale, purchaseFilter, saleFilter } from 'types/valuations'
import type { RealEstate } from 'types/assets/realEstate'
import type { Address } from 'types/commonTypes'
import type { AddressErrors } from 'components/forms/AddressField'
import type { Valuations } from 'types/valuations'
import type { FieldErrors } from 'utils/FormValidation'

import 'styles/forms.css'
import 'styles/buttons.css'
import 'styles/tables.css'
import 'components/assets/real-estate/RealEstateForm.css'

type RealEstateErrors = {
  +name: FieldErrors,
  +address: AddressErrors,
  +notes: FieldErrors,
  +purchaseDate: FieldErrors,
  +purchaseAmount: FieldErrors,
  +saleDate: FieldErrors,
  +saleAmount: FieldErrors,
  +valuations: FieldErrors,
}
const RealEstateErrorsDefaults = {
  name: [],
  address: AddressErrorsDefaults,
  notes: [],
  purchaseDate: [],
  purchaseAmount: [],
  saleDate: [],
  saleAmount: [],
  valuations: [],
}

type Props = {
  +realEstate?: RealEstate,
  +onSubmit?: (realEstate: RealEstate) => void
}

type State = {
  +realEstate: RealEstate,
  +purchase?: { date: string, amount: number },
  +sale?: { date: string, amount: number },
  +errors?: RealEstateErrors,
  +allErrorsShown: boolean,
  +focusedInput: string
}

class RealEstateForm extends React.Component {
  props: Props
  state: State
  handleChange: (fieldName: string, value: string) => void
  handleFocus: (fieldName: string) => void
  handleAddressChange: (address: Address) => void
  handleValuationsChange: (valuations: Valuations) => void
  handlePurchaseChange: (purchase: { +date?: string, +amount?: number }) => void
  handleSaleChange: (sale: { +date?: string, +amount?: number }) => void
  handleSubmit: (event: Event) => boolean

  constructor(props: Props) {
    super(props)
    this.state = {
      realEstate: RealEstateDefaults,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.realEstate) {
      const realEstate = props.realEstate
      const valuations = props.realEstate.valuations
        ? props.realEstate.valuations
        : []
      this.state = {
        ...this.state,
        realEstate,
        purchase: _.pick(valuations.find(purchaseFilter), [ 'date', 'amount' ]),
        sale: _.pick(valuations.find(saleFilter), [ 'date', 'amount' ]),
        errors: RealEstateForm.validate(realEstate),
        allErrorsShown: true,
      }
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleAddressChange = this.handleAddressChange.bind(this)
    this.handleValuationsChange = this.handleValuationsChange.bind(this)
    this.handlePurchaseChange = this.handlePurchaseChange.bind(this)
    this.handleSaleChange = this.handleSaleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName: string, value: string): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, [fieldName]: value }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleFocus(fieldName: string): void {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleAddressChange(address: Address): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, address }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleValuationsChange(valuations: Valuations): void {
    this.setState(prevState => {
      let purchaseVal = valuations.find(v => v.type === 'purchase')
      let saleVal = valuations.find(v => v.type === 'sale')
      let updatedValuations = valuations
      if (purchaseVal) {
        purchaseVal = _.pick(purchaseVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithPurchase(valuations, purchaseVal)
      }
      if (saleVal) {
        saleVal = _.pick(saleVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithSale(valuations, saleVal)
      }
      const realEstate = { ...prevState.realEstate, valuations: updatedValuations }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseOrSaleChange(purchaseOrSale: { +date?: string, +amount?: number }, type: 'purchase'|'sale'): void {
    this.setState(prevState => {
      const updatedPurchaseOrSale = { ...prevState[type], ...purchaseOrSale }
      const updatedValuations = updateValuationsWithPurchaseOrSale(prevState.realEstate.valuations, updatedPurchaseOrSale, type)
      const realEstate = {
        ...prevState.realEstate,
        valuations: updatedValuations,
        [ type === 'purchase' ? 'purchaseDate' : 'saleDate' ]: updatedPurchaseOrSale.date,
      }
      return {
        ...prevState,
        realEstate,
        [ type === 'purchase' ? 'purchase' : 'sale' ]: updatedPurchaseOrSale,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseChange(purchase: { +date?: string, +amount?: number }): void {
    return this.handlePurchaseOrSaleChange(purchase, 'purchase')
  }

  handleSaleChange(sale: { +date?: string, +amount?: number }): void {
    return this.handlePurchaseOrSaleChange(sale, 'sale')
  }

  handleSubmit(event: Event): boolean {
    event.preventDefault()
    this.setState(
      () => ({ errors: RealEstateForm.validate(this.state.realEstate) }),
      () => {
        const { errors, realEstate } = this.state
        if (typeof errors !== 'undefined' && Validations.areErrorsPresent(errors)) {
          const firstErrorFieldName = RealEstateForm.findFirstErrorFieldName(errors)
          firstErrorFieldName
            ? this.setState(() => ({ allErrorsShown: true, focusedInput: firstErrorFieldName }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) { this.props.onSubmit(realEstate) }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.realEstate) {
      this.setState(() => ({ realEstate: nextProps.realEstate }))
    }
  }

  render() {
    const { realEstate, purchase, sale, allErrorsShown, focusedInput } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField = typeof realEstate.id === 'string'
      ? <HiddenField name='id' value={realEstate.id} />
      : null
    const minValuationDate = purchase && purchase.date
      ? moment(purchase.date, DateStorageFormat).add(1, 'd') : undefined
    const maxValuationDate = sale && sale.date
      ? moment(sale.date, DateStorageFormat).subtract(1, 'd') : undefined

    return (
      <form className='real-estate-form form form-aligned' onSubmit={this.handleSubmit}>
        <fieldset>
          <legend>General details</legend>

          {idField}
          <InputField name='name' type='text' label='Name' value={realEstate.name}
            errors={errors.name} forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
          <AddressField name='address' address={realEstate.address} errors={errors.address}
            onChange={this.handleAddressChange}
            onFocus={this.handleFocus} focus={focusedInput} />
          <TextAreaField name='notes' label='Notes' value={realEstate.notes}
            onChange={value => this.handleChange('notes', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Purchase</legend>
          <DateField name='purchaseDate' label='Purchase date'
            value={purchase ? purchase.date : undefined} errors={errors.purchaseDate}
            forceErrorDisplay={allErrorsShown}
            maxDate={maxValuationDate}
            onChange={value => this.handlePurchaseChange({ date: value })}
            onFocus={this.handleFocus} focus={focusedInput} />
          <InputField name='purchaseAmount' label='Purchase amount' type='number' min='1'
            value={purchase && purchase.amount ? purchase.amount.toString() : undefined}
            errors={errors.purchaseAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handlePurchaseChange({ amount: parseInt(value) })}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Sale</legend>
          <DateField name='saleDate' label='Sale date'
            value={sale ? sale.date : undefined} errors={errors.saleDate}
            forceErrorDisplay={allErrorsShown}
            minDate={minValuationDate}
            onChange={value => this.handleSaleChange({ date: value })}
            onFocus={this.handleFocus} focus={focusedInput} />
          <InputField name='saleAmount' label='Sale amount' type='number' min='1'
            value={sale && sale.amount ? sale.amount.toString() : undefined}
            errors={errors.saleAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleSaleChange({ amount: parseInt(value) })}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Value</legend>
          <ValuationsInput name='valuations' valuations={realEstate.valuations}
            minDate={minValuationDate} maxDate={maxValuationDate}
            errors={errors.valuations} forceErrorDisplay={allErrorsShown}
            onChange={this.handleValuationsChange} onFocus={this.handleFocus} />
        </fieldset>
        <fieldset>
          <legend>Mortgage</legend>
          <button type='submit' className='button button-primary'>Submit</button>
        </fieldset>
      </form>
    )
  }

  static validate(realEstate: RealEstate): RealEstateErrors {
    const name = []
      .concat(Validations.required(realEstate.name))
      .concat(Validations.minLength(realEstate.name, 3))
      .concat(Validations.maxLength(realEstate.name, 100))
    const address = AddressField.validate(realEstate.address)
    const purchaseDate = []
      .concat(Validations.required(realEstate.purchaseDate))
    const valuations = []
      .concat(ValuationsInput.validate(realEstate.valuations))
      .concat(RealEstateForm.purchaseValuationPresent(realEstate))
      .concat(RealEstateForm.valuationsWithinPurchaseAndSale(realEstate))

    return {
      ...RealEstateErrorsDefaults,
      name,
      address,
      purchaseDate,
      valuations,
    }
  }

  static purchaseValuationPresent(realEstate: RealEstate): FieldErrors {
    return Validations.validate(
      realEstate,
      re => re.valuations.some(valuation => valuation.type === 'purchase'),
      'There must be a purchase valuation'
    )
  }

  static valuationsWithinPurchaseAndSale(realEstate: RealEstate): FieldErrors {
    const purchaseVal = realEstate.valuations.find(v => v.type === 'purchase')
    const saleVal = realEstate.valuations.find(v => v.type === 'sale')
    if (!purchaseVal) { return [] }

    return Validations.validate(
      realEstate,
      re => re.valuations.every(valuation =>
        [ 'purchase', 'sale' ].includes(valuation.type) || (saleVal
          ? valuation.date > purchaseVal.date && valuation.date < saleVal.date
          : valuation.date > purchaseVal.date)
      ),
      'Valuations must all be between the purchase and sale dates'
    )
  }

  static findFirstErrorFieldName(realEstateErrors: RealEstateErrors): string|null {
    return (function seek(realEstateErrors, parentComponentName) {
      let match = null
      _.some(realEstateErrors, (errors, fieldName) => {
        if (errors instanceof Array) {
          if (errors.length > 0) {
            match = parentComponentName ? parentComponentName + '-' + fieldName : fieldName
            return true
          } else return false
        } else if (errors instanceof Object) {
          match = seek(errors, fieldName)
          return (match !== null)
        }
      })
      return match
    })(realEstateErrors)
  }
}

export default RealEstateForm
