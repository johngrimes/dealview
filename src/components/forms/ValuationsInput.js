// @flow

import React from 'react'
import DatePicker from 'react-datepicker'
import moment from 'moment'

import InputField from '../../components/forms/InputField.js'

import './ValuationsInput.css'
import 'react-datepicker/dist/react-datepicker-cssmodules.css'

type Valuation = {
  date: string,
  amount: number,
  note: string
}

export type Valuations = Array<Valuation>

export const ValuationsEmpty = []

type Props = {
  name: string,
  value: Valuations,
  notifyChange?: (valuations: Valuations) => void
}

type State = {
  value: Valuations
}

class ValuationsInput extends React.Component {
  props: Props
  state: State
  handleChange: (i: number, field: string, value: string|number) => void
  handleAddValuation: (event: Event) => true

  constructor(props: Props) {
    super(props)
    this.state = { value: this.props.value }

    this.handleChange = (i, field, value) => {
      const newState = this.state
      newState.value[i][field] = value
      this.setState(newState)
      if (this.props.notifyChange) this.props.notifyChange(newState.value)
    }

    this.handleAddValuation = (event) => {
      const newState = this.state
      newState.value.push({ date: '1983-06-21', amount: 1, note: 'Some note' })
      this.setState(newState)
      if (this.props.notifyChange) this.props.notifyChange(newState.value)
      return true
    }
  }

  render() {
    const valuations = this.state.value.map((v, i) => {
      return <tr key={i}>
        <td><DatePicker name={`valuations-date-${i}`} dateFormat='YYYY-MM-DD' selected={moment(v.date, 'YYYY-MM-DD')}
          onChange={(value) => this.handleChange(i, 'date', value.format('YYYY-MM-DD'))} /></td>
        <td><InputField name={`valuations-amount-${i}`} type='number' value={v.amount.toString()}
          notifyChange={(value) => this.handleChange(i, 'amount', parseInt(value, 10))} /></td>
        <td><InputField name={`valuations-note-${i}`} type='text' value={v.note}
          notifyChange={(value) => this.handleChange(i, 'note', value)} /></td>
      </tr>
    })
    const header = valuations.length > 0 ? <thead>
      <tr>
        <th>Date</th>
        <th>Amount</th>
        <th>Note</th>
      </tr>
    </thead> : null
    return (
      <div className='valuations-input control-group'>
        <table className='table'>
          {header}
          <tbody>{valuations}</tbody>
        </table>
        <button className='button add-valuation-button' type='button' onClick={this.handleAddValuation}>+&nbsp;&nbsp;Add</button>
      </div>
    )
  }
}

export default ValuationsInput
