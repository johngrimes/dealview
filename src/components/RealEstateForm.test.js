import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'

import RealEstateForm from './RealEstateForm.js'
import ValuationsInput from './ValuationsInput.js'
import {
  validRealEstate1,
  validRealEstate2,
} from '../data/fixtures/realEstate.js'
import { DateStorageFormat } from '../data/commonTypes.js'
import { RealEstateDefaults } from '../data/realEstate.js'

describe('RealEstateForm', () => {
  it('should render', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    expect(wrapper).toMatchSnapshot()
    const wrapper2 = shallow(<RealEstateForm realEstate={validRealEstate2} />)
    expect(wrapper2).toMatchSnapshot()
  })

  it('should pass valuations to ValuationsInput component', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    expect(wrapper.find(ValuationsInput).prop('valuations')).toBe(
      validRealEstate1.valuations
    )
  })

  it('should pass minimum date to ValuationsInput component when there is a purchase', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    const minDate = wrapper.find(ValuationsInput).prop('minDate')
    const expectedMinDate = moment(
      validRealEstate1.purchaseDate,
      DateStorageFormat
    ).add(1, 'd')
    expect(minDate.isSame(expectedMinDate)).toBe(true)
  })

  it('should pass maximum date to ValuationsInput component when there is a sale', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate2} />)
    const maxDate = wrapper.find(ValuationsInput).prop('maxDate')
    const expectedMaxDate = moment(
      validRealEstate2.saleDate,
      DateStorageFormat
    ).subtract(1, 'd')
    expect(maxDate.isSame(expectedMaxDate)).toBe(true)
  })

  it('should not pass minimum date to ValuationsInput when purchase has no date', () => {
    const realEstate = {
      ...validRealEstate1,
      valuations: [{ amount: 165000, note: 'Purchase price', type: 'purchase' }],
    }
    const wrapper = shallow(<RealEstateForm realEstate={realEstate} />)
    const minDate = wrapper.find(ValuationsInput).prop('minDate')
    expect(minDate).toBe(undefined)
  })

  it('should not pass maximum date to ValuationsInput when sale has no date', () => {
    const realEstate = {
      ...validRealEstate1,
      valuations: [{ amount: 340000, note: 'Sale price', type: 'sale' }],
    }
    const wrapper = shallow(<RealEstateForm realEstate={realEstate} />)
    const maxDate = wrapper.find(ValuationsInput).prop('maxDate')
    expect(maxDate).toBe(undefined)
  })

  it('should focus the first field with errors upon submission', () => {
    const wrapper = shallow(<RealEstateForm />)
    wrapper.find({ name: 'name' }).prop('onChange')('rover')
    wrapper.find({ name: 'address' }).prop('onChange')({
      line1: '',
      line2: '',
      line3: 'a',
      locality: '',
      state: '',
      postcode: '',
    })
    wrapper.find('form').simulate('submit', { preventDefault() {} })
    expect(wrapper.state('focusedInput')).toEqual('address-line3')
  })

  it('should update state when a field is focused', () => {
    const wrapper = shallow(<RealEstateForm />)
    wrapper.find({ name: 'name' }).prop('onFocus')('name')
    expect(wrapper.state('focusedInput')).toBe('name')
  })

  it('should show all errors upon submission', () => {
    const wrapper = shallow(<RealEstateForm />)
    wrapper.find('form').simulate('submit', { preventDefault() {} })
    expect(wrapper.state('allErrorsShown')).toEqual(true)
  })
})

describe('validation', () => {
  it('should require a purchase validation', () => {
    const realEstate = { ...validRealEstate1 }
    realEstate.valuations = realEstate.valuations.filter(
      v => v.type !== 'purchase'
    )
    const wrapper = shallow(<RealEstateForm realEstate={realEstate} />)
    expect(wrapper.find(ValuationsInput).prop('errors')).toMatchSnapshot()
  })

  it('should require valuations to be between purchase and sale dates', () => {
    const realEstate = { ...validRealEstate1 }
    realEstate.valuations = realEstate.valuations.concat([
      {
        date: '2014-09-09',
        amount: 460000,
        note: 'Some other valuation',
        type: 'none',
      },
    ])
    const wrapper = shallow(<RealEstateForm realEstate={realEstate} />)
    expect(wrapper.find(ValuationsInput).prop('errors')).toMatchSnapshot()
  })
})

describe('purchase and sale', () => {
  it('should update state when purchase date is changed', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    wrapper.find({ name: 'purchaseDate' }).prop('onChange')('2016-05-14')
    expect(wrapper.state()).toMatchSnapshot()
  })

  it('should update state when purchase amount is changed', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    wrapper.find({ name: 'purchaseAmount' }).prop('onChange')(350000)
    expect(wrapper.state()).toMatchSnapshot()
  })

  it('should update state when sale date is changed', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    wrapper.find({ name: 'saleDate' }).prop('onChange')('2016-05-14')
    expect(wrapper.state()).toMatchSnapshot()
  })

  it('should update state when sale amount is changed', () => {
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate1} />)
    wrapper.find({ name: 'saleAmount' }).prop('onChange')(350000)
    expect(wrapper.state()).toMatchSnapshot()
  })

  it('should not add empty purchase and sale upon edit of a valuation', () => {
    const realEstate = RealEstateDefaults
    const wrapper = shallow(<RealEstateForm realEstate={realEstate} />)
    wrapper.find(ValuationsInput).prop('onChange')([{ amount: 1 }])
    expect(wrapper.state('realEstate').valuations).toHaveLength(1)
  })

  it('should set minimum on sale date when purchase is present', () => {
    const expected = moment('2005-05-10', DateStorageFormat).add(1, 'd')
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate2} />)
    expect(
      wrapper.find({ name: 'saleDate' }).prop('minDate').isSame(expected)
    ).toBe(true)
  })

  it('should set maximum on purchase date when sale is present', () => {
    const expected = moment('2017-01-01', DateStorageFormat).subtract(1, 'd')
    const wrapper = shallow(<RealEstateForm realEstate={validRealEstate2} />)
    expect(
      wrapper.find({ name: 'purchaseDate' }).prop('maxDate').isSame(expected)
    ).toBe(true)
  })
})
