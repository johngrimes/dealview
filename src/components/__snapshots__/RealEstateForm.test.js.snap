// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RealEstateForm should render 1`] = `
<form
  className="real-estate-form form form-aligned"
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      General details
    </legend>
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Name"
      name="name"
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="5 Sunrise St"
    />
    <AddressField
      address={
        Object {
          "line1": "5 Sunrise St",
          "line2": "",
          "line3": "",
          "locality": "Ashgrove",
          "postcode": "4060",
          "state": "QLD",
        }
      }
      errors={
        Object {
          "line1": Array [],
          "line2": Array [],
          "line3": Array [],
        }
      }
      focus="name"
      name="address"
      onChange={[Function]}
      onFocus={[Function]}
    />
    <TextAreaField
      focus="name"
      label="Notes"
      name="notes"
      onChange={[Function]}
      onFocus={[Function]}
      value="Principal place of residence"
    />
  </fieldset>
  <fieldset>
    <legend>
      Purchase
    </legend>
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Purchase date"
      name="purchaseDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2014-09-10"
    />
    <CurrencyField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Purchase amount"
      name="purchaseAmount"
      onChange={[Function]}
      onFocus={[Function]}
      value="520000"
    />
  </fieldset>
  <fieldset>
    <legend>
      Sale
    </legend>
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Sale date"
      minDate={"2014-09-10T14:00:00.000Z"}
      name="saleDate"
      onChange={[Function]}
      onFocus={[Function]}
    />
    <CurrencyField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Sale amount"
      name="saleAmount"
      onChange={[Function]}
      onFocus={[Function]}
    />
  </fieldset>
  <fieldset>
    <legend>
      Value
    </legend>
    <ValuationsInput
      errors={Array []}
      forceErrorDisplay={true}
      minDate={"2014-09-10T14:00:00.000Z"}
      name="valuations"
      onChange={[Function]}
      onFocus={[Function]}
      valuations={
        Array [
          Object {
            "amount": 520000,
            "date": "2014-09-10",
            "note": "Purchase price",
            "type": "purchase",
          },
          Object {
            "amount": 705000,
            "date": "2016-04-15",
            "note": "Formal valuation",
            "type": "none",
          },
        ]
      }
    />
  </fieldset>
  <fieldset>
    <legend>
      Mortgage
    </legend>
    <button
      className="button button-primary"
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;

exports[`RealEstateForm should render 2`] = `
<form
  className="real-estate-form form form-aligned"
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      General details
    </legend>
    <InputField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Name"
      name="name"
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="38/8 Briggs Rd"
    />
    <AddressField
      address={
        Object {
          "line1": "Unit 38, Capertree Gardens",
          "line2": "8 Briggs Rd",
          "line3": "",
          "locality": "Springwood",
          "postcode": "4127",
          "state": "QLD",
        }
      }
      errors={
        Object {
          "line1": Array [],
          "line2": Array [],
          "line3": Array [],
        }
      }
      focus="name"
      name="address"
      onChange={[Function]}
      onFocus={[Function]}
    />
    <TextAreaField
      focus="name"
      label="Notes"
      name="notes"
      onChange={[Function]}
      onFocus={[Function]}
      value="Investment property"
    />
  </fieldset>
  <fieldset>
    <legend>
      Purchase
    </legend>
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Purchase date"
      maxDate={"2016-12-30T14:00:00.000Z"}
      name="purchaseDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2005-05-10"
    />
    <CurrencyField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Purchase amount"
      name="purchaseAmount"
      onChange={[Function]}
      onFocus={[Function]}
      value="165000"
    />
  </fieldset>
  <fieldset>
    <legend>
      Sale
    </legend>
    <DateField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Sale date"
      minDate={"2005-05-10T14:00:00.000Z"}
      name="saleDate"
      onChange={[Function]}
      onFocus={[Function]}
      value="2017-01-01"
    />
    <CurrencyField
      errors={Array []}
      focus="name"
      forceErrorDisplay={true}
      label="Sale amount"
      name="saleAmount"
      onChange={[Function]}
      onFocus={[Function]}
      value="340000"
    />
  </fieldset>
  <fieldset>
    <legend>
      Value
    </legend>
    <ValuationsInput
      errors={Array []}
      forceErrorDisplay={true}
      maxDate={"2016-12-30T14:00:00.000Z"}
      minDate={"2005-05-10T14:00:00.000Z"}
      name="valuations"
      onChange={[Function]}
      onFocus={[Function]}
      valuations={
        Array [
          Object {
            "amount": 165000,
            "date": "2005-05-10",
            "note": "Purchase price",
            "type": "purchase",
          },
          Object {
            "amount": 260000,
            "date": "2015-09-10",
            "note": "Formal valuation",
            "type": "none",
          },
          Object {
            "amount": 340000,
            "date": "2017-01-01",
            "note": "Sale price",
            "type": "sale",
          },
        ]
      }
    />
  </fieldset>
  <fieldset>
    <legend>
      Mortgage
    </legend>
    <button
      className="button button-primary"
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;

exports[`purchase and sale should update state when purchase amount is changed 1`] = `
Object {
  "allErrorsShown": true,
  "errors": Object {
    "address": Object {
      "line1": Array [],
      "line2": Array [],
      "line3": Array [],
    },
    "name": Array [],
    "notes": Array [],
    "purchaseAmount": Array [],
    "purchaseDate": Array [],
    "saleAmount": Array [],
    "saleDate": Array [],
    "valuations": Array [],
  },
  "focusedInput": "name",
  "purchase": Object {
    "amount": 350000,
    "date": "2014-09-10",
  },
  "realEstate": Object {
    "address": Object {
      "line1": "5 Sunrise St",
      "line2": "",
      "line3": "",
      "locality": "Ashgrove",
      "postcode": "4060",
      "state": "QLD",
    },
    "name": "5 Sunrise St",
    "notes": "Principal place of residence",
    "purchaseDate": "2014-09-10",
    "valuations": Array [
      Object {
        "amount": 705000,
        "date": "2016-04-15",
        "note": "Formal valuation",
        "type": "none",
      },
      Object {
        "amount": 350000,
        "date": "2014-09-10",
        "note": "Purchase price",
        "type": "purchase",
      },
    ],
  },
  "sale": Object {},
}
`;

exports[`purchase and sale should update state when purchase date is changed 1`] = `
Object {
  "allErrorsShown": true,
  "errors": Object {
    "address": Object {
      "line1": Array [],
      "line2": Array [],
      "line3": Array [],
    },
    "name": Array [],
    "notes": Array [],
    "purchaseAmount": Array [],
    "purchaseDate": Array [],
    "saleAmount": Array [],
    "saleDate": Array [],
    "valuations": Array [
      "Valuations must all be between the purchase and sale dates",
    ],
  },
  "focusedInput": "name",
  "purchase": Object {
    "amount": 520000,
    "date": "2016-05-14",
  },
  "realEstate": Object {
    "address": Object {
      "line1": "5 Sunrise St",
      "line2": "",
      "line3": "",
      "locality": "Ashgrove",
      "postcode": "4060",
      "state": "QLD",
    },
    "name": "5 Sunrise St",
    "notes": "Principal place of residence",
    "purchaseDate": "2016-05-14",
    "valuations": Array [
      Object {
        "amount": 705000,
        "date": "2016-04-15",
        "note": "Formal valuation",
        "type": "none",
      },
      Object {
        "amount": 520000,
        "date": "2016-05-14",
        "note": "Purchase price",
        "type": "purchase",
      },
    ],
  },
  "sale": Object {},
}
`;

exports[`purchase and sale should update state when sale amount is changed 1`] = `
Object {
  "allErrorsShown": true,
  "errors": Object {
    "address": Object {
      "line1": Array [],
      "line2": Array [],
      "line3": Array [],
    },
    "name": Array [],
    "notes": Array [],
    "purchaseAmount": Array [],
    "purchaseDate": Array [],
    "saleAmount": Array [],
    "saleDate": Array [],
    "valuations": Array [
      "Valuations must all have dates",
      "Valuations must all be between the purchase and sale dates",
    ],
  },
  "focusedInput": "name",
  "purchase": Object {
    "amount": 520000,
    "date": "2014-09-10",
  },
  "realEstate": Object {
    "address": Object {
      "line1": "5 Sunrise St",
      "line2": "",
      "line3": "",
      "locality": "Ashgrove",
      "postcode": "4060",
      "state": "QLD",
    },
    "name": "5 Sunrise St",
    "notes": "Principal place of residence",
    "purchaseDate": "2014-09-10",
    "saleDate": undefined,
    "valuations": Array [
      Object {
        "amount": 520000,
        "date": "2014-09-10",
        "note": "Purchase price",
        "type": "purchase",
      },
      Object {
        "amount": 705000,
        "date": "2016-04-15",
        "note": "Formal valuation",
        "type": "none",
      },
      Object {
        "amount": 350000,
        "note": "Sale amount",
        "type": "sale",
      },
    ],
  },
  "sale": Object {
    "amount": 350000,
  },
}
`;

exports[`purchase and sale should update state when sale date is changed 1`] = `
Object {
  "allErrorsShown": true,
  "errors": Object {
    "address": Object {
      "line1": Array [],
      "line2": Array [],
      "line3": Array [],
    },
    "name": Array [],
    "notes": Array [],
    "purchaseAmount": Array [],
    "purchaseDate": Array [],
    "saleAmount": Array [],
    "saleDate": Array [],
    "valuations": Array [
      "Valuations must all have amounts",
    ],
  },
  "focusedInput": "name",
  "purchase": Object {
    "amount": 520000,
    "date": "2014-09-10",
  },
  "realEstate": Object {
    "address": Object {
      "line1": "5 Sunrise St",
      "line2": "",
      "line3": "",
      "locality": "Ashgrove",
      "postcode": "4060",
      "state": "QLD",
    },
    "name": "5 Sunrise St",
    "notes": "Principal place of residence",
    "purchaseDate": "2014-09-10",
    "saleDate": "2016-05-14",
    "valuations": Array [
      Object {
        "amount": 520000,
        "date": "2014-09-10",
        "note": "Purchase price",
        "type": "purchase",
      },
      Object {
        "amount": 705000,
        "date": "2016-04-15",
        "note": "Formal valuation",
        "type": "none",
      },
      Object {
        "date": "2016-05-14",
        "note": "Sale amount",
        "type": "sale",
      },
    ],
  },
  "sale": Object {
    "date": "2016-05-14",
  },
}
`;

exports[`validation should require a purchase validation 1`] = `
Array [
  "There must be a purchase valuation",
]
`;

exports[`validation should require valuations to be between purchase and sale dates 1`] = `
Array [
  "Valuations must all be between the purchase and sale dates",
]
`;
